format/ComputeIR.fbs

```diff
@@ -0,0 +1,521 @@
+table Aggregate {
+  input:TableExpr;
+  aggregate_exprs:[ArrayExpr] (required);
+
+  /// Expressions to use as group keys. If not provided, then the aggregate
+  /// operation yields a table with a single row.
+  group_exprs:[ArrayExpr];
+
+  /// A filter condition for the aggregation which may include aggregate
+  /// functions.
+  having:[ArrayExpr];
+}
+
+/// Select up to the indicated number of rows from the input expression based
+/// on the first-emitted rows when evaluating the input expression. Generally,
+/// no particular order is guaranteed unless combined with a sort expression.
+table Limit {
+  input:TableExpr;
+
+  /// Number of logical rows to select from input.
+  limit:long;
+}
+
+enum RelationalJoinType : int {
+  INNER = 0,
+  LEFT = 1,
+  RIGHT = 2,
+  OUTER = 3,
+  SEMI = 4,
+  ANTI = 5,
+}
+
+/// A standard relational / SQL-style equijoin.
+///
+/// Providing no left/right columns produces the cross product of the two
+/// tables.
+table EqualityJoin {
+  // TODO: complete and document
+  type:RelationalJoinType = INNER;
+
+  left:TableExpr (required);
+  right:TableExpr (required);
+
+  /// The columns from the left expression to use when joining
+  left_columns:[ColumnReference] (required);
+
+  /// The columns from the right expression to use when joining. Must have the
+  /// same length as left_columns.
+  ///
+  /// If omitted, the names provided in left_columns must match the names in
+  /// the right expression.
+  right_columns:[ColumnReference];
+}
+
+/// A relational non-equijoin containing expressions which may include
+/// inequality or range conditions.
+table NonEqualityJoin {
+  // TODO: complete and document
+  type:RelationalJoinType = INNER;
+  left:TableExpr (required);
+  right:TableExpr (required);
+  left_exprs:[ArrayExpr];
+  right_exprs:[ArrayExpr];
+}
+
+/// A temporal join type
+///
+/// TODO: complete and document
+table AsOfJoin {
+  left:TableExpr (required);
+  right:TableExpr;
+
+  /// The column in the left expression to use for data ordering to determine
+  /// the "as of" time.
+  left_asof:ColumnReference (required);
+
+  /// The column in the right expression to use for data ordering to determine
+  /// the "as of" time. If the column name is the same as left_asof, may be
+  /// omitted.
+  right_asof:ColumnReference;
+
+  /// TODO: Define means of providing time deltas in this IR.
+  tolerance:Literal;
+
+  /// If true, the
+  allow_equal:bool = true;
+}
+
+/// An extension of as-of join which allows applying an aggregate function to
+/// the data falling within the indicated time interval.
+///
+/// TODO: Define semantics of "identity" window where all elements of window
+/// become a List<T> element in the result.
+table WindowJoin {
+  // TODO
+}
+
+/// An expression to use for sorting. The key expression determines the values
+/// to be used for ordering the table's rows.
+table SortKey {
+  key:ArrayExpr (required);
+  ascending:bool = true;
```

Comment:
This is one of the API choices I really dislike about pandas, and think SQL did well: a specific language construct for indicating sort order. I personally find it really annoying to have to compute a logical negation in my head to get descending behavior, and I think about code in pandas like `df.sort_values([("a", False), ("b", True)])` which actually requires me to know what the heck the bools correspond to there, as opposed to an enum like `Order.Asc`, which reads clear.

IMO we should do the same and have an `enum SortOrder { Ascending, Descending }` and replace the `ascending` with an `order` field.